archive.go

```diff
@@ -2,157 +2,88 @@ package k6deps
 
 import (
 	"archive/tar"
+	"bytes"
 	"encoding/json"
 	"errors"
 	"io"
-	"os"
 	"path/filepath"
-	"strings"
-
-	"github.com/grafana/k6pack"
+	"slices"
 )
 
-//nolint:forbidigo
-func loadMetadata(dir string, opts *Options) error {
-	var meta archiveMetadata
-
-	data, err := os.ReadFile(filepath.Join(filepath.Clean(dir), "metadata.json"))
-	if err != nil {
-		return err
-	}
-
-	if err = json.Unmarshal(data, &meta); err != nil {
-		return err
-	}
-
-	opts.Manifest.Ignore = true // no manifest (yet) in archive
-
-	opts.Script.Name = filepath.Join(
-		dir,
-		"file",
-		filepath.FromSlash(strings.TrimPrefix(meta.Filename, "file:///")),
-	)
-
-	if value, found := meta.Env[EnvDependencies]; found {
-		opts.Env.Name = EnvDependencies
-		opts.Env.Contents = []byte(value)
-	} else {
-		opts.Env.Ignore = true
-	}
-
-	contents, err := os.ReadFile(opts.Script.Name)
-	if err != nil {
-		return err
-	}
-
-	script, _, err := k6pack.Pack(string(contents), &k6pack.Options{Filename: opts.Script.Name})
-	if err != nil {
-		return err
-	}
-
-	opts.Script.Contents = script
-
-	return nil
-}
-
 type archiveMetadata struct {
 	Filename string            `json:"filename"`
 	Env      map[string]string `json:"env"`
 }
 
 const maxFileSize = 1024 * 1024 * 10 // 10M
 
-//nolint:forbidigo
-func extractArchive(dir string, input io.Reader) error {
+func processArchive(input io.Reader) (analyzer, error) {
 	reader := tar.NewReader(input)
 
+	analyzers := make([]analyzer, 0)
+
 	for {
 		header, err := reader.Next()
 
 		switch {
-		case err == io.EOF:
-			return nil
+		case errors.Is(err, io.EOF):
+			return mergeAnalyzers(analyzers...), nil
 		case err != nil:
-			return err
+			return nil, err
 		case header == nil:
 			continue
 		}
 
-		target := filepath.Join(dir, filepath.Clean(filepath.FromSlash(header.Name)))
-
-		switch header.Typeflag {
-		case tar.TypeDir:
-			if err := os.MkdirAll(target, 0o750); err != nil {
-				return err
-			}
+		if header.Typeflag != tar.TypeReg || !shouldProcess(header.Name) {
+			continue
+		}
 
-		case tar.TypeReg:
-			if shouldSkip(target) {
-				continue
-			}
+		content := &bytes.Buffer{}
+		if _, err := io.CopyN(content, reader, maxFileSize); err != nil && !errors.Is(err, io.EOF) {
```

Comment:
It's a corner case, but still, I'm thinking isn't that right for the files that are bigger than 10 MB :thinking: Shoudld we maybe write a warning log in that case.

If the other suggestion about `analizeMetadata` will be accepted, we could probably move these lines closer to the `scriptAnalyzer`

And perhaps not for that PR, but it's worth investigation if we can instead of copying `content.Bytes()` just pass around readers, in other words  make possible to `scriptAnalyzer` also work with buffers

Reply:
> It's a corner case, but still, I'm thinking isn't that right for the files that are bigger than 10 MB 🤔 Shoudld we maybe write a warning log in that case.

Not sure what you refer to in "isn't that right". Could you please elaborate?

Reply:
So my understanding that this line will copy at maximum 10MB, silently ignoring the rest, and I'm questioning if that's a right way to do. Yes, the risks that we processing bigger files where module usage located after 10MB of data is low, but still

Reply:
Sure, like I said, both my options are non-blocking, feel free to merge this as it is and continue in follow-up PRs :+1: